module/bdev/aio/bdev_aio.c

```diff
@@ -120,46 +120,135 @@ static struct spdk_bdev_module aio_if = {
 
 SPDK_BDEV_MODULE_REGISTER(aio, &aio_if)
 
+#define ERROR(rc) (((rc) > 0) ? -(rc) : -EIO)
+
 static int
 bdev_aio_open(struct file_disk *disk)
 {
 	int fd;
 
-	fd = open(disk->filename, O_RDWR | O_DIRECT);
+	fd = open(disk->filename, (O_RDWR | O_DIRECT));
 	if (fd < 0) {
 		/* Try without O_DIRECT for non-disk files */
 		fd = open(disk->filename, O_RDWR);
 		if (fd < 0) {
-			SPDK_ERRLOG("open() failed (file:%s), errno %d: %s\n",
-				    disk->filename, errno, spdk_strerror(errno));
+			int rc = errno;
+
+			SPDK_ERRLOG("open(%s) failed, errno %d: %s\n",
+				    disk->filename, rc, spdk_strerror(rc));
 			disk->fd = -1;
-			return -1;
+			return ERROR(rc);
 		}
 	}
 
 	disk->fd = fd;
-
 	return 0;
 }
 
+/* zero region of a (sparse) file by creating a hole */
 static int
-bdev_aio_close(struct file_disk *disk)
+aio_range_fallocate_sparse(int fd, uint64_t range[])
+{
+	return fallocate(fd, (FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE), range[0], range[1]);
+}
+
+/* zero region of a file by writing zeros */
+static int
+aio_range_fallocate_zero(int fd, uint64_t range[])
 {
+	return fallocate(fd, (FALLOC_FL_ZERO_RANGE | FALLOC_FL_KEEP_SIZE), range[0], range[1]);
+}
+
+/* unmap region of a device */
+static int
+aio_range_unmap_discard(int fd, uint64_t range[])
+{
+	return ioctl(fd, BLKDISCARD, range);
+}
+
+/* zero region of a device */
+static int
+aio_range_unmap_zero(int fd, uint64_t range[])
+{
+	return ioctl(fd, BLKZEROOUT, range);
+}
+
+static int
+set_aio_range_functions(struct file_disk *disk)
+{
+	struct stat st;
 	int rc;
 
-	if (disk->fd == -1) {
+	if (fstat(disk->fd, &st) < 0) {
+		rc = errno;
+		SPDK_ERRLOG("fstat(%s) failed, errno %d: %s\n",
+			    disk->filename, rc, spdk_strerror(rc));
+		disk->unmap = NULL;
+		disk->zero = NULL;
+		return ERROR(rc);
+	}
+
+	if (S_ISBLK(st.st_mode)) {
+		int discard = 0;
+
+		if (ioctl(disk->fd, BLKDISCARDZEROES, &discard) < 0) {
```

Comment:
I'm beginning to think that it might not be a good idea to try an use discard at all. However, for now, have made change that failure of the ioctl() is effectively treated as if it had returned 0. And it seems that it will always return 0 on newer kernels anyway.